src/c.c

```diff
@@ -146,6 +146,10 @@ bool needs_vec_c_fallback(SEXP xs) {
     return false;
   }
 
+  if (ptype != R_NilValue && !equal_object(r_class(x), r_class(ptype))) {
```

Comment:
I guess the class check is good enough? I think thats the only thing that was mildly surprising here

Reply:
hmm I think we don't want to make that check when `ptype` is `NULL`, otherwise we'd return false without continuing the fallback.